[MRG] Fixed kwargs issue breaking raw_to_bids

a-hurst opened the pull request:
PR Description
--------------

Fixes #455 by making `allow_maxshield` an actual argument to `_read_raw`, preventing it from getting passed via kwargs to reader functions that don't support it.

Note that this doesn't fix `raw_to_bids` completely for EDF and BDF files. The writer code has a check that raises an exception if data has been preloaded, and "preload=True" is explicitly set for the ".edf" and ".bdf" readers in `_read_raw`:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/mne_bids/commands/mne_bids_raw_to_bids.py", line 81, in <module>
    run()
  File "/usr/local/lib/python3.7/site-packages/mne_bids/commands/mne_bids_raw_to_bids.py", line 77, in run
    verbose=True)
  File "/usr/local/lib/python3.7/site-packages/mne_bids/write.py", line 1128, in write_raw_bids
    raise ValueError('The data should not be preloaded.')
ValueError: The data should not be preloaded.
```

The relevant chunks of code are here:

https://github.com/mne-tools/mne-bids/blob/237b3badab5fe8c28b64bf68b398780d1b53fa7d/mne_bids/read.py#L56-L60

https://github.com/mne-tools/mne-bids/blob/237b3badab5fe8c28b64bf68b398780d1b53fa7d/mne_bids/write.py#L944-L945

Looking at the Git blame, that TODO comment in read.py two years old, and setting "preload=False" fixes the issue completely for me, suggesting that the mentioned MNE feature has since been added.

Is it fine for me to drop the `preload` argument from those readers as part of this PR as well, or would that accidentally break compatibility with older MNE versions or something?


Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [x] All comments resolved
- [x] This is not your own PR
- [x] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [x] PR description includes phrase "closes <#issue-number>"
- [x] Commit history does not contain any merge commits

jasmainak commented:
> Fixes #455 by making allow_maxshield an actual argument to _read_raw, preventing it from getting passed via kwargs to reader functions that don't support it.

why would this ever happen though? Your fix will now work only for `allow_maxshield` but not for other mismatches between file formats and their readers.

> The writer code has a check that raises an exception if data has been preloaded, 

yes this is on purpose

> s it fine for me to drop the preload argument from those readers as part of this PR as well, or would that accidentally break compatibility with older MNE versions or something?

I think you can address it in this PR too since we're already relying on latest version of MNE, right @sappelhoff ?

a-hurst commented:
> why would this ever happen though? Your fix will now work only for `allow_maxshield` but not for other mismatches between file formats and their readers.

It looks like the functionality was added in [this commit](https://github.com/mne-tools/mne-bids/commit/e1547dd4035232544e591bc20f36cd81f4ccc50d), which replaced a specific `allow_maxshield` flag with the current generic `**kwargs`. The only example of it I see getting used in the code for a non-`allow_maxshield` reason is in this test, where it's getting used to pass a `clean_names = False` argument to a reader:

https://github.com/mne-tools/mne-bids/blob/237b3badab5fe8c28b64bf68b398780d1b53fa7d/mne_bids/tests/test_write.py#L580-L584

I've only been using mne-bids for the past two days, so I really don't have that strong a sense of whether this is useful or not.

jasmainak commented:
I see, you're trying to fix the CLI. I would start first by writing a test so we know for sure that the bug is fixed.

a-hurst commented:
Okay, added an EDF file test to the CLI tests (hopefully won't make things too much slower) and removed the `preload=True` flag for ".bdf" and ".edf" files. Hopefully this looks okay!

jasmainak commented:
can you update `whats_new.rst` and set the PR to [MRG] when you're ready.

@hoechenberger do you have a moment to check this PR?

a-hurst commented:
Alright, think I've got everything in order on my end. I think the codecov failure is just because I removed the EDF/BDF specific lines in read.py, which *were* covered by the tests, meaning there's proportionally slightly less coverage overall?

agramfort commented:
I don't think it's the right fix. I would change mne_bids_raw_to_bids.py and typically do not set allow_maxshield to True if opt.raw_fname.endswith('.fif'). That's all

here you can to read_raw things that are specific to one format.

sappelhoff commented:
> I don't think it's the right fix. I would change mne_bids_raw_to_bids.py and typically do not set allow_maxshield to True if opt.raw_fname.endswith('.fif'). That's all
here you can to read_raw things that are specific to one format.

but wouldn't your argument suggest, that we should deal differently with `electrode=None, hsp=None, hpi=None, config=None` as well?

Something like:

original code:

https://github.com/mne-tools/mne-bids/blob/237b3badab5fe8c28b64bf68b398780d1b53fa7d/mne_bids/commands/mne_bids_raw_to_bids.py#L72

first change:

```diff
-    parser.add_option('--allow_maxshield', dest='allow_maxshield',
-                     help="whether to allow non maxfiltered data (BOOLEAN)",
-                      action='store_true')
+    parser.add_option('--allow_maxshield', dest='allow_maxshield',
+                     help="whether to allow non maxfiltered data (BOOLEAN)",
+                     )
```

then adjust:

```diff
-    raw = _read_raw(opt.raw_fname, hpi=opt.hpi, electrode=opt.electrode,
-                    hsp=opt.hsp, config=opt.config,
-                    allow_maxshield=allow_maxshield)
+    extra_params = dict(hpi=opt.hpi, electrode=opt.electrode, hsp=opt.hsp,
+                       config=opt.config, allow_maxshield=opt.allow_maxshield)
+    extra_params = {key: val, for key, val in extra_params.items() if val is not None}
+    raw = _read_raw(opt.raw_fname, **extra_params)
```

--> and of course then also adjust https://github.com/mne-tools/mne-bids/blob/237b3badab5fe8c28b64bf68b398780d1b53fa7d/mne_bids/read.py#L33


???

agramfort commented:
ok fait enough @sappelhoff we should follow the current logic. So LGTM

jasmainak commented:
The fix is good enough for me as @sappelhoff suggested that the "proper fix" would be more complicated and it's not clear if it's worth it. Please feel free to merge once the merge commit is gone. We could probably already merge since we squash and merge? I'm not sure.

a-hurst commented:
Alright, I fixed the order of the arguments as @sappelhoff suggested and rebased the PR (think I did this right). Hopefully now everything's good to go!
